### PR TITLE
Scaffold burn on validator set admission

### PR DESCRIPTION
**Pull request by lpgeiger**
- [x] implements a flat 1 GAS burn to join the validator set, at epoch boundary.
- [x] user can set burn preferences.
- [x] optionally sends to burn wallet, or send to the community wallets weighted average.
- [x] use the out of band deposits to community wallets to compute weighted average.
- [x] decrease over time the relative weight of the donations (newer donations more relevant).
- [x] optional setting on accounts to track cumulative balance deposited, needed for community wallets.
- [x] functional tests

### CODE REVIEW

**Comment by zmanian** [comment]
Why does this change improve OL?

**Review by jamesmeijers — COMMENTED** [review]
I'm not sure I really see the point of this change, but I don't necessarily have anything against it.

**Comment by lpgeiger** [comment]
@zmanian @jamesmeijers  In combination with https://github.com/OLSF/libra/pull/588, this would obviate the need for donations using the autopay mechanism. More here: https://docs.google.com/document/d/1uFrZBeb5_oSFTuSnaoaNsavpeWDXN-4Ewn0yoR0RqLI/edit#heading=h.bnws9kc42b3c

File: language/stdlib/modules/0L/Burn.move [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  public fun epoch_start_burn(vm: &signer, payer: address, value: u64) acquires DepositInfo, BurnPreference {
+    if (exists<BurnPreference>(payer)) {
+      if (borrow_global<BurnPreference>(payer).is_burn) {
+        return burn(vm, payer, value)
+      }
+    };
+    send(vm, payer, value);
+  }
+
+  fun burn(vm: &signer, payer: address, value: u64) {
```
[the remark this answers is not in the document]
Comment (lpgeiger) [reply language/stdlib/modules/0L/Burn.move:88]: Yes that would be ideal. The issue is that the Libra burning mechanism is quite cumbersome, requiring a pre-burn, and then approval from a treasury. When we first looked at burns a year ago, sending to 0xDEADDEAD was the provisional solution. So yes, we should move toward something with more assurances, but I think the semantics are sufficient for the time being. Thoughts?

File: language/stdlib/modules/0L/AutoPay.move [hunk 2]
```diff
@@ -125,6 +126,33 @@ address 0x1 {
...
+        let account_addr = Vector::borrow<address>(account_list, account_idx);
+        // Obtain the account balance
+        // let account_bal = LibraAccount::balance<GAS>(*account_addr);
+        // Go through all payments for this account and pay 
+        let payments = &mut borrow_global_mut<Data>(*account_addr).payments;
+        let payments_len = Vector::length<Payment>(payments);
+        let payments_idx = 0;
+        while (payments_idx < payments_len) {
+          let payment = Vector::borrow_mut<Payment>(payments, payments_idx);
+          Vector::push_back<address>(&mut payee_vec, payment.payee);
```
[the remark this answers is not in the document]
Comment (lpgeiger) [reply language/stdlib/modules/0L/AutoPay.move:147]: yeah good catch, that's a bug.

File: language/stdlib/modules/0L/Burn.move (hunk 1 above)
[the remark this answers is not in the document]
Comment (jamesmeijers) [reply language/stdlib/modules/0L/Burn.move:88]: Sounds fine to me.

File: language/stdlib/modules/0L/AutoPay.move (hunk 2 above)
[the remark this answers is not in the document]
Comment (lpgeiger) [reply language/stdlib/modules/0L/AutoPay.move]: ok fixed

**Review by earthwindfirewater — APPROVED** [review]
Looks good

**Pull request closed without merging**